controllers/metal3machinetemplate_controller.go

```diff
@@ -0,0 +1,115 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	capm3 "github.com/metal3-io/cluster-api-provider-metal3/api/v1alpha4"
+	"github.com/metal3-io/cluster-api-provider-metal3/baremetal"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	templateControllerName = "Metal3MachineTemplate-controller"
+)
+
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=metal3machinestemplates,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=metal3machinestemplates/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=metal3machines,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=metal3machines/status,verbs=get
+
+// Metal3MachineTemplateReconciler reconciles a Metal3MachineTemplate object
+type Metal3MachineTemplateReconciler struct {
+	Client         client.Client
+	ManagerFactory baremetal.ManagerFactoryInterface
+	Log            logr.Logger
+}
+
+// Reconcile handles Metal3MachineTemplate events
+func (r *Metal3MachineTemplateReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, rerr error) {
+	ctx := context.Background()
+	m3templateLog := r.Log.WithValues(templateControllerName).WithValues("metal3-machine-template", req.NamespacedName)
+
+	// Fetch the Metal3MachineTemplate instance.
+	metal3MachineTemplate := &capm3.Metal3MachineTemplate{}
+
+	if err := r.Client.Get(ctx, req.NamespacedName, metal3MachineTemplate); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, errors.Wrap(err, "unable to fetch Metal3MachineTemplate")
+	}
+
+	helper, err := patch.NewHelper(metal3MachineTemplate, r.Client)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to init patch helper")
+	}
+
+	// Always patch metal3MachineTemplate exiting this function so we can persist any metal3MachineTemplate changes.
+	defer func() {
+		err := helper.Patch(ctx, metal3MachineTemplate)
+		if err != nil {
+			m3templateLog.Info("failed to Patch Metal3MachineTemplate")
+		}
+	}()
+
+	// Fetch the Metal3Machine instance.
+	m3machine := &capm3.Metal3Machine{}
+
+	// Fetch the Metal3MachineList
+	m3machinelist := &capm3.Metal3MachineList{}
+
+	// Create a helper for managing a Metal3MachineTemplate.
+	templateMgr, err := r.ManagerFactory.NewMachineTemplateManager(metal3MachineTemplate, m3machine, m3machinelist, m3templateLog)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to create helper for managing the templateMgr")
+	}
+
+	// Handle non-deleted machines
+	return r.reconcileNormal(ctx, templateMgr)
```

Comment:
Why do we need to have to reconcile function `reconcile` and `reconcileNormal`? Do you plan to have another reconcile function for another case?

Reply:
Thanks @namnx228 . Currently we don't have `reconcileDelete()`, because we right don't do anything with M3MT with deletion timestamps. And, why `reconcileNormal` is because I just used the same reconciliation flow as in any other controller code. I can move all the logic under `reconcile` if you think it would be better, TBH I don't have strong opinion on that.

Reply:
No, I also don't against the idea to have `reconcileNormal`. Let keep it like this.

Reply:
Okay, will do that. Thanks